### PR TITLE
Shibboleth dev4

### PR DESCRIPTION
**Pull request by Flusspferd123**
*************  WORK IN PROGRESS ********************
### Simple Approach with button “Hardcoded Shibb”
User clicks button “Hardcoded Shibb”. This button’s href property is set to the url the Shibboleth SP (Service Provider) uses as an entry point for the SSO authentication. Href url gets `dspace.hostname/shibboleth` as target parameter.  (E.g. href = `https://dspace.hostname/Shibboleth.sso/Login?target=https://dspace.hostname/shibboleth`) . The target parameter is the url the SP redirects the user to after a successful shibboleth authentication. 
The Angular route `/shibboleth` points to a component which dispatches a “`GetJWTafterShibbLoginAction`” in its` ngOnInit()` method.  This triggers (via an ngrx-effect) a REST-API POST call to “`spring-rest/api/authn/login`”. If the Shibboleth Authentication process was successful this call triggers a backend response carrying a JWT token and the user gets logged-in.
When testing this approach make sure you change the href attribute of the “Hardcodedshibb” button in “`log-in.component.html`” to `https://dspace.hostname/Shibboleth.sso/Login?target=https://dspace.hostname/shibboleth` in order to make this work with your Shibboleth IdentityProvider and Service Provider. ( `dspace.hostname` = as in  backend config file "dspace.cfg".)  Do not change the following parts of href: "`/Shibboleth.sso/Login?target=`" and "`/shibboleth`" at the end of href.
### What’s the problem with this approach? 
It works but on startup the Angular frontend should check if Shibboleth Authentication is enabled at the backend. The button triggering the shibboleth Authentication should only be visible if Shibboleth is enabled. 
### Approach “Is Shibboleth enabled at the backend?”
On startup of the angular app a `“RETRIEVE_AUTH_METHODS`” Action is dispatched.  This triggers a POST call to “`spring-rest/api/authn/login`”. The backend returns a response with status 405 and the following header:
 `WWW-Authenticate →shibboleth realm="DSpace REST API", location="/Shibboleth.sso/Login?target=https://dspace.hostname", password realm="DSpace REST API"`
This response is intercepted in “`auth.interceptor.ts`” where the header is parsed and the ssoLoginURL is built. This url looks exactly as the url  used for the “HardCoded shibb” button. 
A `“RETRIEVE_AUTH_METHODS_SUCCESS`” Action stores the shibbolethURL in the ngrx-store. (`core.auth. ssoLoginUrl)` The “`log-in.component.ts`” is subscribed to this ssoLoginUrl via  “`this.store.pipe(select(getSSOLoginUrl))`” and displays the button “Shibboleth” if a ssoLoginUrl is stored in the ngrx-store.
### What’s the problem with the “Is Shibboleth enabled at the backend” approach?
It works when using “`yarn run watch`”. When using “`yarn start`” or other production build commands the following error occurs: 
```
Error in SSR, serving for direct CSR
Error: Can't set headers after they are sent.
    at SendStream.headersAlreadySent (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\send\index.js:390:13)
    at SendStream.send (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\send\index.js:618:10)
    at onstat (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\send\index.js:730:10)
    at ZoneDelegate.invokeTask (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\zone.js\dist\zone-node.js:421:31)
    at Zone.runTask (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\zone.js\dist\zone-node.js:188:47)
    at ZoneTask.invokeTask (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\zone.js\dist\zone-node.js:496:34)
    at ZoneTask.invoke (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\zone.js\dist\zone-node.js:485:48)
    at data.args.(anonymous function) (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\zone.js\dist\zone-node.js:1317:25)
    at FSReqWrap.oncomplete (fs.js:153:5)
```

Some feedback on why this is the case would be highly appreciated.

### CODE REVIEW

**Comment by paulo-graca** [comment]
Thank you @Flusspferd123 for adding this! This it's very useful for us and we would like to participate in helping you bringing this to DSpace. But before my colleague Fernando take part on a feedback, I would like to understand if you have really have more than 1000 files changed to support Shibboleth, or if they are also related with other changes? In the case they are related with other aspects in DSpace, is it possible to have a separate and specific branch for the Shibboleth PR? This would be very useful to facilitate the reviewing process.

**Comment by AlexanderS** [comment]
I think all the new `*.js` files and all `*.map` files should not be added. This would make 1580 changed files less.

**Comment by atarix83** [comment]
@Flusspferd123 Before to start a review I think you need rebase your branche because as @paulo-graca said there are too many file changes

**Comment by Flusspferd123** [comment]
hi everbody
sorry for the delay - i was on holyday. Alexander was right: the js and map files were not meant to be part of this pull request as these are files generated by the intelij javscript debugger. Please have a look at this PR now: Is this is workable for you?
Please also note: This is work in progress and is only a basis for further discussion on how to implement the shibboleth  login

**Comment by atarix83** [comment]
@Flusspferd123 
I had a first look to the code an this is my consideration. Compared to the first implementation of the rest server, now it returns back a list of available authentication methods with `WWW-Authenticate` header. 
So we could now use this header to refactor authentication module to allow to retrieve all authentication methods on client side. Something similar of what is done on rest [login.html](https://github.com/DSpace/DSpace/blob/d896fed7a71da0b4ad123ca541367dac9436ccb0/dspace-server-webapp/src/main/webapp/login.html#L103)

This should be our scenario with shibboleth : 
```
WWW-Authenticate →shibboleth realm="DSpace REST API", location="https://dspace7.4science.cloud/Shibboleth.sso/Login?target=https%3A%2F%2Fdspace7.4science.cloud", password realm="DSpace REST API"
```
so this means that on our angular client app we can provide password authentication and shibboleth autentication. 
This could be achieve with these steps:
- add available auth methods on [AuthState](https://github.com/Flusspferd123/dspace-angular/blob/d018017c79c48f2e58a9310e0da064e77352b76e/src/app/core/auth/auth.reducer.ts#L22) 
- retrieve available auth methods and location by `WWW-Authenticate` header. You could just improve and use [AuthService#retrieveAuthMethods](https://github.com/Flusspferd123/dspace-angular/blob/d018017c79c48f2e58a9310e0da064e77352b76e/src/app/core/auth/auth.service.ts#L225)
- Change [log-in.component](https://github.com/Flusspferd123/dspace-angular/blob/d018017c79c48f2e58a9310e0da064e77352b76e/src/app/shared/log-in/log-in.component.html) to allow to show only auth methods available. This means we should create a new component for any method supported (password and shibboleth in this scenario).
We could use `Injector` and custom `decorator` to make this dynamic and to easily add new auth method in the future (this approch is used for example to render [submssion sections](https://github.com/Flusspferd123/dspace-angular/blob/d018017c79c48f2e58a9310e0da064e77352b76e/src/app/submission/sections/container/section-container.component.html#L44) )

Let me know if you need any further clarification.

**Comment by Flusspferd123** [comment]
I made the changes suggested by Guiseppe. CSS Styling and Test are still missing. But the original problem persits:
It works when using “yarn run watch”. When using “yarn start” or other production build commands the following error occurs:
`Error in SSR, serving for direct CSR
Error: Can't set headers after they are sent.
    at SendStream.headersAlreadySent (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\send\index.js:390:13)
    at SendStream.send (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\send\index.js:618:10)
    at onstat (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\send\index.js:730:10)
    at ZoneDelegate.invokeTask (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\zone.js\dist\zone-node.js:421:31)
    at Zone.runTask (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\zone.js\dist\zone-node.js:188:47)
    at ZoneTask.invokeTask (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\zone.js\dist\zone-node.js:496:34)
    at ZoneTask.invoke (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\zone.js\dist\zone-node.js:485:48)
    at data.args.(anonymous function) (C:\Users\gruber\IdeaProjects\DSpace7-Angular\node_modules\zone.js\dist\zone-node.js:1317:25)
    at FSReqWrap.oncomplete (fs.js:153:5)`

Some feedback on whats causing this error would be highly appreciated

File: src/app/shared/log-in/log-in-container/log-in-container.component.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      providers: [
+        {provide: 'shibbolethUrlProvider', useFactory: () => (this.shibbolethUrl), deps: []},
+        // if other authentication methods need further data to work add a provider here e.g
+        // {provide: 'otherDataProvider', useFactory: () => (this.otherData), deps: []},
+      ],
+      parent: this.injector
+    });
+
+    this.dynamicLoginMethods = this.store.select(getAuthenticationMethods).pipe(
+      map(((authMethods) => authMethods.map((authMethod) => {
```
Comment (atarix83) [inline src/app/shared/log-in/log-in-container/log-in-container.component.ts]: I'd avoid to have a fixed switch case, but I'd use a decorator to dinamically add login method.
[Here](https://github.com/DSpace/dspace-angular/blob/master/src/app/submission/sections/sections-decorator.ts) an example and [here](https://github.com/DSpace/dspace-angular/blob/master/src/app/submission/sections/form/section-form.component.ts#L40) how is used

File: src/app/+login-page/shibbolethTargetPage/shibboleth.component.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import {Store} from '@ngrx/store';
+import {CoreState} from '../../core/core.reducers';
+import {Observable, of} from 'rxjs';
+
+@Component({
+  selector: 'ds-shibboleth-page',
+  templateUrl: './shibboleth.component.html',
+  styleUrls: ['./shibboleth.component.scss']
+})
+export class ShibbolethComponent implements OnInit {
```
Comment (atarix83) [inline src/app/+login-page/shibbolethTargetPage/shibboleth.component.ts]: Please could you explain to me the reason for this component? Is this necessary?

File: src/app/core/auth/auth.actions.ts [hunk 3]
```diff
@@ -308,6 +325,39 @@ export class ResetAuthenticationMessagesAction implements Action {
...
+ * Check if token is already present upon initial load.
+ * @class CheckAuthenticationTokenAction
+ * @implements {Action}
+ */
+export class RetrieveAuthMethodsSuccessAction implements Action {
+  public type: string = AuthActionTypes.RETRIEVE_AUTH_METHODS_SUCCESS;
+  payload: string;
+
+  constructor(location: string) {
+    this.payload = location;
```
Comment (atarix83) [inline src/app/core/auth/auth.actions.ts]: payload should be an array of `AuthMethodModel`, isn't it?

File: src/app/core/auth/auth.actions.ts [hunk 4]
```diff
@@ -48,10 +52,19 @@ export class AuthenticateAction implements Action {
...
+    this.payload = {email, password};
   }
 }
 
+/**
+ * GetJWTafterShibbLoginAction.
+ * @class GetJWTafterShibbLoginAction
+ * @implements {Action}
+ */
+export class GetJWTafterShibbLoginAction implements Action {
```
Comment (atarix83) [inline src/app/core/auth/auth.actions.ts]: How is used this new action?

File: src/app/core/auth/auth.interceptor.ts [hunk 5]
```diff
@@ -49,17 +52,65 @@ export class AuthInterceptor implements HttpInterceptor {
...
+    // tslint:disable-next-line:forin
+    for (const i in parts) {
+      const part: string = parts[i].trim();
+      if (part.includes('realm')) {
+        const methodName = part.split(' ')[0];
+        const authMethod: AuthMethodModel = new AuthMethodModel(methodName);
+        // check if the authentication method is  shibboleth
+        // if so the next part is the shibboleth location
+        // e.g part i: shibboleth realm="DSpace REST API", part i+1:  location="/Shibboleth.sso/Login?target=https://serverUrl"
+        if (methodName.includes('shibboleth')) {
```
Comment (atarix83) [inline src/app/core/auth/auth.interceptor.ts]: is it possible to have a generic function to retrieve authentication method? this should avoid to manually add other check like this for every new authentication method that we would like to implement in the future
I think [this is](https://github.com/DSpace/DSpace/blob/d896fed7a71da0b4ad123ca541367dac9436ccb0/dspace-server-webapp/src/main/webapp/login.html#L103) a good example on how to implement it

File: src/app/core/auth/auth.service.ts [hunk 6]
```diff
@@ -1,27 +1,30 @@
...
-import { AuthTokenInfo, TOKENITEM } from './models/auth-token-info.model';
-import { isEmpty, isNotEmpty, isNotNull, isNotUndefined } from '../../shared/empty.util';
-import { CookieService } from '../../shared/services/cookie.service';
-import { getAuthenticationToken, getRedirectUrl, isAuthenticated, isTokenRefreshing } from './selectors';
-import { AppState, routerStateSelector } from '../../app.reducer';
-import { ResetAuthenticationMessagesAction, SetRedirectUrlAction } from './auth.actions';
-import { NativeWindowRef, NativeWindowService } from '../../shared/services/window.service';
-import { Base64EncodeUrl } from '../../shared/utils/encode-decode.util';
-import { RemoteDataBuildService } from '../cache/builders/remote-data-build.service';
+import {Inject, Injectable, Optional} from '@angular/core';
```
Comment (atarix83) [inline src/app/core/auth/auth.service.ts]: please leave space between curly brackets in imports

File: src/app/core/auth/models/auth-method.model.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+      }
+
+      default: {
+        break;
+      }
+    }
+  }
+}
+
+export enum AuthMethodConstants {
```
Comment (atarix83) [inline src/app/core/auth/models/auth-method.model.ts]: please use string enum so you can easily map with auth method name 
```
export enum AuthMethodType {
  Password = 'password',
  Shibboleth = 'shibboleth',
  Ldap = 'ldap',
  Ip = 'ip'
}
```

File: src/app/shared/log-in/password/log-in.component.html [hunk 8]
```diff
@@ -17,9 +18,15 @@
...
+  <div *ngIf="(error | async) && hasError" class="alert alert-danger" role="alert"
+       @fadeOut>{{ (error | async) | translate }}</div>
+  <div *ngIf="(message | async) && hasMessage" class="alert alert-info" role="alert"
+       @fadeOut>{{ (message | async) | translate }}</div>
+
+  <button class="btn btn-lg btn-primary btn-block mt-3" type="submit"
+          [disabled]="!form.valid">{{"login.form.submit" | translate}}</button>
+
+
   <div class="dropdown-divider"></div>
```
Comment (atarix83) [inline src/app/shared/log-in/password/log-in.component.html]: the password login component shoul contain only form and submit button, I should move all the rest to the login container

File: src/app/shared/log-in/x509/dynamic-x509.component.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+ */
+import {Component, OnInit} from '@angular/core';
+
+@Component({
+  selector: 'ds-dynamic-x509',
+  templateUrl: './dynamic-x509.component.html',
+  styleUrls: ['./dynamic-x509.component.scss'],
+
+})
+export class DynamicX509Component {
```
Comment (atarix83) [inline src/app/shared/log-in/x509/dynamic-x509.component.ts]: I would not implement this component in this PR

File: src/app/shared/log-in/ldap/dynamic-ldap.component.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+ */
+import {Component, OnInit} from '@angular/core';
+
+@Component({
+  selector: 'ds-dynamic-ldap',
+  templateUrl: './dynamic-ldap.component.html',
+  styleUrls: ['./dynamic-ldap.component.scss'],
+
+})
+export class DynamicLdapComponent {
```
Comment (atarix83) [inline src/app/shared/log-in/ldap/dynamic-ldap.component.ts]: I would not implement this component in this PR

File: src/app/shared/log-in/ip/dynamic-ip.component.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+ */
+import {Component, OnInit} from '@angular/core';
+
+@Component({
+  selector: 'ds-dynamic-ip',
+  templateUrl: './dynamic-ip.component.html',
+  styleUrls: ['./dynamic-ip.component.scss'],
+
+})
+export class DynamicIpComponent {
```
Comment (atarix83) [inline src/app/shared/log-in/ip/dynamic-ip.component.ts]: I would not implement this component in this PR

File: src/app/shared/log-in/x509/dynamic-x509.component.ts (hunk 9 above)
Comment (Flusspferd123) [reply src/app/shared/log-in/x509/dynamic-x509.component.ts]: Deleted the x509 component form the PR

File: src/app/shared/log-in/ldap/dynamic-ldap.component.ts (hunk 10 above)
Comment (Flusspferd123) [reply src/app/shared/log-in/ldap/dynamic-ldap.component.ts]: Deleted the LDAP component.

File: src/app/shared/log-in/ip/dynamic-ip.component.ts (hunk 11 above)
Comment (Flusspferd123) [reply src/app/shared/log-in/ip/dynamic-ip.component.ts]: Deleted the IP component.

File: src/app/core/auth/auth.actions.ts (hunk 3 above)
Comment (Flusspferd123) [reply src/app/core/auth/auth.actions.ts]: Changed the payload type to AuthMethodModel[]

File: src/app/core/auth/models/auth-method.model.ts (hunk 7 above)
Comment (Flusspferd123) [reply src/app/core/auth/models/auth-method.model.ts]: Changed it as suggested

File: src/app/shared/log-in/log-in-container/log-in-container.component.ts (hunk 1 above)
Comment (Flusspferd123) [reply src/app/shared/log-in/log-in-container/log-in-container.component.ts]: Changed it to the approach with decorator and ngComonentOutlet directive

File: src/app/shared/log-in/password/log-in.component.html (hunk 8 above)
Comment (Flusspferd123) [reply src/app/shared/log-in/password/log-in.component.html]: Did you mean to make the log-in.component a presentational component?

File: src/app/core/auth/auth.service.ts (hunk 6 above)
Comment (Flusspferd123) [reply src/app/core/auth/auth.service.ts]: Now i found the intelij setting that you already showed me at OR2019. So future code will have spaces

File: src/app/core/auth/auth.interceptor.ts (hunk 5 above)
Comment (Flusspferd123) [reply src/app/core/auth/auth.interceptor.ts]: Tryed to make it more generic....please check it.

File: src/app/+login-page/shibbolethTargetPage/shibboleth.component.ts (hunk 2 above)
Comment (Flusspferd123) [reply src/app/+login-page/shibbolethTargetPage/shibboleth.component.ts]: The target parameter of the shibboleth URL  is the URL the SP redirects the user to after a successful shibboleth authentication. In this implementation the target is ..../shibboleth.
This Angular route (/shibboleth) points to this component which dispatches a “GetJWTafterShibbLoginAction” in its ngOnInit() method. This triggers (via an ngrx-effect) a REST-API POST call to “spring-rest/api/authn/login”. If the Shibboleth Authentication process was successful this call triggers a backend response carrying a JWT token and the user gets logged-in.

**Comment by Flusspferd123** [comment]
Reopening message

File: src/app/core/auth/auth.interceptor.ts [hunk 12]
```diff
@@ -49,17 +52,72 @@ export class AuthInterceptor implements HttpInterceptor {
...
+        location = this.parseLocation(location)
+        authMethodModel.location = location;
+        console.log('location: ', location);
+        authMethodModels.push(authMethodModel);
+      }
+    }
+    return authMethodModels;
+  }
+
+  private makeAuthStatusObject(authenticated: boolean, accessToken?: string, error?: string, httpHeaders?: HttpHeaders,): AuthStatus {
```
[the remark this answers is not in the document]
Comment (Flusspferd123) [reply src/app/core/auth/auth.interceptor.ts]: Deleted the comma

File: src/app/shared/log-in/methods/shibboleth/dynamic-shibboleth.component.html [hunk 13]
```diff
@@ -0,0 +1,11 @@
+<form *ngIf="!(loading | async) && !(isAuthenticated | async)" class="form-login px-4 py-3" (ngSubmit)="submit()"
```
[the remark this answers is not in the document]
Comment (Flusspferd123) [reply src/app/shared/log-in/methods/shibboleth/dynamic-shibboleth.component.html]: Removed the form

File: src/app/shared/log-in/authMethods.component.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+   * True if the authentication is loading.
+   * @type {boolean}
+   */
+  public loading: Observable<boolean>;
+
+  constructor( private store: Store<CoreState>) {
+  }
+
+  ngOnInit(): void {
+    this.authMethodData =  this.authMethodData = this.store.select(getAuthenticationMethods);
```
[the remark this answers is not in the document]
Comment (Flusspferd123) [reply src/app/shared/log-in/authMethods.component.ts]: Changed it as suggested

File: src/app/shared/log-in/authMethods-type.ts [hunk 15]
```diff
@@ -0,0 +1,7 @@
+export enum AuthMethodType {
```
[the remark this answers is not in the document]
Comment (Flusspferd123) [reply src/app/shared/log-in/authMethods-type.ts]: moved the file to methods folder

File: src/app/shared/log-in/authMethods-decorator.ts [hunk 16]
```diff
@@ -0,0 +1,16 @@
+import { AuthMethodType } from './authMethods-type';
```
[the remark this answers is not in the document]
Comment (Flusspferd123) [reply src/app/shared/log-in/authMethods-decorator.ts]: moved the fiile to methods folder

File: src/app/shared/log-in/authMethods.component.ts [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+import { select, Store } from '@ngrx/store';
+import { getAuthenticationMethods, isAuthenticated, isAuthenticationLoading } from '../../core/auth/selectors';
+import { CoreState } from '../../core/core.reducers';
+
+@Component({
+  selector: 'ds-auth-methods',
+  templateUrl: './authMethods.component.html',
+  styleUrls: ['./authMethods.component.scss']
+})
+export class AuthMethodsComponent implements OnInit {
```
[the remark this answers is not in the document]
Comment (Flusspferd123) [reply src/app/shared/log-in/authMethods.component.ts]: Changed class and file name

File: src/app/shared/log-in/methods/shibboleth/dynamic-shibboleth.component.ts [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+    this.loading = this.store.pipe(select(isAuthenticationLoading));
+
+  }
+
+  submit() {
+    console.log('submit() was called');
+    this.store.dispatch(new StartShibbolethAuthenticationAction(this.authMethodModel));
+    this.host = 'fis.tiss.tuwien.ac.at';
+    // https://dspace.hostname/Shibboleth.sso/Login?target=https://dspace.hostname/shibboleth
+    window.location.href = 'https://' + this.host + '/Shibboleth.sso/Login?target=https://' + this.host + '/shibboleth';
```
[the remark this answers is not in the document]
Comment (Flusspferd123) [reply src/app/shared/log-in/methods/shibboleth/dynamic-shibboleth.component.ts]: Now it gets injected

**Comment by paulo-graca** [comment]
Hi @Flusspferd123 , any reason for why I can't see on ds-auth-methods component? I didn't dig enough but, authMethodName is not defined. My configurations seem to be ok because I have both authentication methods on my server side. Am I missing something?

2019_10_03_17_11_16_Sign_in_HAL_Browser

**Pull request merged**